system/jlib/jsecrets.cpp

```diff
@@ -1029,6 +1046,130 @@ extern jlib_decl bool getSecretValue(StringBuffer & result, const char *category
     return true;
 }
 
+//---------------------------------------------------------------------------------------------------------------------
+
+class CSecret final : public CInterfaceOf<ISecret>
+{
+public:
+    CSecret(const char *_category, const char * _name, const char * _vaultId, const char * _version, const IPropertyTree * _secret)
+    : category(_category), name(_name), vaultId(_vaultId), version(_version), secret(_secret)
+    {
+        updateHash();
+    }
+
+    virtual const IPropertyTree * getTree() const;
+
+    virtual bool getKeyValue(MemoryBuffer & result, const char * key) const
+    {
+        CriticalBlock block(secretCs);
+        checkStale();
+        return getSecretKeyValue(result, secret, key);
+    }
+    virtual bool getKeyValue(StringBuffer & result, const char * key) const
+    {
+        CriticalBlock block(secretCs);
+        checkStale();
+        return getSecretKeyValue(result, secret, key);
+    }
+    virtual bool isStale() const
+    {
+        return secret && hasCacheExpired(secret);
+    }
+    virtual unsigned getVersion() const
+    {
+        return secretHash;
+    }
+
+protected:
+    void checkStale() const;
+    void updateHash() const;
+
+protected:
+    StringAttr category;
+    StringAttr name;
+    StringAttr vaultId;
+    StringAttr version;
+    mutable CriticalSection secretCs;
+    mutable Linked<const IPropertyTree> secret;
+    mutable unsigned secretHash = 0;
+};
+
+
+const IPropertyTree * CSecret::getTree() const
+{
+    CriticalBlock block(secretCs);
+    checkStale();
+    return LINK(secret);
+}
+
+void CSecret::checkStale() const
+{
+    if (isStale())
+    {
+        //MORE: This could block or fail - in roxie especially it would be better to return the old value
+        try
+        {
+            secret.setown(getSecretTree(category, name, vaultId, version));
```

Comment:
@ghalliday I wonder if getSecretTree() itself shouldn't fail if the cache can't be updated?  If cache expiration itself is basically just a way of keeping the secret fresh.

It's different then say the certificate itself expiring which is separate and tied to the content becoming invalid.